File: Recursion/GFG/Medium/BSTtoSortedList/BSTtoSortedList.cpp
// https://www.geeksforgeeks.org/problems/flatten-bst-to-sorted-list--111950/1
#include <bits/stdc++.h>
using namespace std;
class Node
{
public:
    int data;
    Node *left;
    Node *right;
    Node(int val)
    {
        data = val;
        left = NULL;
        right = NULL;
    }
};
// Main code
Node *flattenBST(Node *root)
{
    if (root)
    {
        Node *left = flattenBST(root->left); // trust that this code gives you the head of left sub tree which is flattened
        root->left = nullptr;
        Node *head = root;
        if (left)
        {
            head = left;
            while (left->right != nullptr)
            {
                left = left->right;
            }
            left->right = root;
        }
        Node *right = flattenBST(root->right);
        root->right = right;
        return head;
    }
    return nullptr;
}
int main(void)
{

    return 0;
}

File: Recursion/GFG/Medium/BSTtoSortedList/BSTtoSortedList.java
// https://www.geeksforgeeks.org/problems/flatten-bst-to-sorted-list--111950/1

public class BSTtoSortedList {
    // Main Code
    static Node flattenBST(Node root) {
        if (root != null) {
            Node left = flattenBST(root.left); // trust that this code gives you the head of left sub tree which is flattened
            root.left = null;
            Node head = root;
            if (left != null) {
                head = left;
                while (left.right != null) {
                    left = left.right;
                }
                left.right = root;
            }
            root.right = flattenBST(root.right);
            return head;
        }
        return null;
    }

    public static void main(String[] args) {

    }
}

class Node {

    int data;
    Node left, right;

    Node(int val) {
        data = val;
        left = right = null;
    }
}
